# LAB 02.03 - Pandas

In [78]:
!wget --no-cache -O init.py -q https://raw.githubusercontent.com/rramosp/ai4eng.v1/main/content/init.py
import init; init.init(force_download=False); init.get_weblink()


In [125]:
from local.lib.rlxmoocapi import submit, session
session.LoginSequence(endpoint=init.endpoint, session_id='2025.2', course_id=init.course_id, lab_id="L02.03", varname="student");

logging in as santiago.herrera6@udea.edu.co... please wait
you are registered in the following sessions of this course

     ['2025.2', 'open']

and you have explicitly requested session '2025.2'.
your grades will be reported in that session

-------------
using course session ai4eng.v1.udea.r3::2025.2
success!! you are logged in
-------------


In [ ]:
import numpy as np
import pandas as pd

## Task 1: Extract data

we have a dataframe of items with a price and, **sometimes**, with an extra column (`margin`).

You will have to complete a function that will **filter** the dataframe selecting the rows:

- whose price is > 100
- **OR** whose margin is >10, if the column `margin` is present

your function must return **A LIST** with the item ids of the selected rows

For instance, with the following dataframe

             price  category
    itemid                  
    56556   108.15       2.0
    73065    83.54       2.0
    36619   114.42       0.0
    73414    82.94       0.0
    13410   115.13       2.0
    66153    91.50       1.0
    77380    85.82       0.0
    73249    95.95       0.0
    11654   100.22       2.0
    11972    77.16       1.0

your must return the following list:

     [56556, 36619, 13410, 11654]
     
But if you get the following dataframe

             price  category  margin
    itemid                          
    39059    98.11       0.0   11.04
    19526    98.11       1.0   11.25
    78176    94.34       1.0   10.51
    50948   102.37       1.0   10.77
    12111    98.07       1.0    8.50
    56191    98.53       1.0   11.65
    38887    91.49       2.0   11.24
    77915   117.30       0.0    8.64
    55010    96.13       0.0    8.95
    45925    98.59       1.0   10.45
    
you must return the following list

    [39059, 19526, 78176, 50948, 56191, 38887, 77915, 45925]

In [43]:
price=np.random.normal(loc=100, scale=10, size=10)
print('Price \n',price)
nmissing = np.random.randint(len(price)//2)+2
print(nmissing,type(nmissing))
a=np.random.permutation(len(price))
print(a)
print(a[nmissing])
price[a[nmissing]] = np.nan
print('New Price \n',price)

print(len(price)//2)

Price 
 [104.64718368 105.79013007  97.97474792  87.68971882 105.3095716
  93.6302334   89.46280985 109.65974501  95.87397414 101.9630231 ]
2 <class 'int'>
[8 4 3 1 5 7 2 0 9 6]
3
New Price 
 [104.64718368 105.79013007  97.97474792          nan 105.3095716
  93.6302334   89.46280985 109.65974501  95.87397414 101.9630231 ]
5


In [87]:
def create_df(missing=False, n=10):
    itemid   = np.random.randint(100000, size=n)+1000
    category = np.random.randint(3, size=n)
    price    = np.round(np.random.normal(loc=100, scale=10, size=n),2)
    margin   = np.round(np.random.normal(loc=10, scale=1, size=n),2)

    if missing:
        nmissing = np.random.randint(len(price)//2)+2
        price[np.random.permutation(len(price))[:nmissing]] = np.nan

    d = pd.DataFrame(np.r_[[price, category, margin]].T, index=itemid, columns=["price", "category", "margin"])
    d.index.name="itemid"
    if np.random.random()>.5:
        d = d[d.columns[:2]]

    return d

In [55]:
np.random.random()

0.47828467493544546

In [88]:
d = create_df()
d

price  category
itemid                  
41086   102.41       0.0
11996   100.24       1.0
18289    99.64       1.0
65126   104.49       2.0
89632   100.01       0.0
13014   104.57       0.0
41717    99.87       0.0
18705    96.32       0.0
80678   106.97       1.0
13951   108.70       1.0

In [92]:
def select_items(df):
    # make sure to make a copy in case you modify the original df
    df = df.copy()
    if 'margin' in df.columns:
      result = df[((df.price>100) | (df.margin>10)) & df.price.notna()].index.tolist()
    else:
      result = df[(df.price>100) & df.price.notna()].index.tolist()
    return result

**manually check your answer**

In [93]:
select_items(d)

[41086, 11996, 65126, 89632, 13014, 80678, 13951]

In [94]:
d

price  category
itemid                  
41086   102.41       0.0
11996   100.24       1.0
18289    99.64       1.0
65126   104.49       2.0
89632   100.01       0.0
13014   104.57       0.0
41717    99.87       0.0
18705    96.32       0.0
80678   106.97       1.0
13951   108.70       1.0

**submit your code**

In [95]:
student.submit_task(globals(), task_id="task_01");

## Task 2: Group statistics

Complete the following function so that it returns a dataframe with the average, max and min **prices** per category.

For instance, for the following dataframe

             price  category  margin
    itemid                          
    17946    93.85       1.0   10.64
    61190    91.72       1.0    9.76
    39639   100.16       1.0   10.67
    17791   110.44       2.0    9.65
    7333    101.05       1.0    9.69
    77362   122.33       0.0   11.14
    92646   108.13       2.0   10.58
    27797    85.52       2.0   10.88
    31746    97.56       0.0    9.75
    12355   101.04       2.0    9.51
    
you should return the following dataframe

                 media  maximo  minimo
    categoria                         
    0         109.9450  122.33  97.56
    1          96.6950  101.05  91.72
    2         101.2825  110.44  85.52
    
observe that your result
- must not be a **multilevel** columnset.
- the column names and the index name must be **exactly** as in the example.
- the **index** must be of type **int**.

In [113]:
media_0 = d[d['category'] == 0]['price'].mean()
max_1 = d[d['category'] == 1]['price'].max()
min_2 = d[d['category'] == 2]['price'].min()
print(media_0, max_1, min_2)

100.636 108.7 104.49


In [122]:
d

price  category
itemid                  
41086   102.41       0.0
11996   100.24       1.0
18289    99.64       1.0
65126   104.49       2.0
89632   100.01       0.0
13014   104.57       0.0
41717    99.87       0.0
18705    96.32       0.0
80678   106.97       1.0
13951   108.70       1.0

In [131]:
def get_stats(df):
    # make sure to make a copy in case you modify the original df
    df=df.copy()
    result=[]
    for i in range(3):

      media = df[df['category'] == i]['price'].mean()
      max = df[df['category'] == i]['price'].max()
      min = df[df['category'] == i]['price'].min()
      result.append({'media': media, 'maximo': max, 'minimo': min})
    data = pd.DataFrame(result)
    data.index.name='categoria'

    return data

**manually check your answer**

In [ ]:
d = create_df()
d

In [123]:
get_stats(d)

media  maximo  minimo
categoria                          
0          100.6360  104.57   96.32
1          103.8875  108.70   99.64
2          104.4900  104.49  104.49

**submit your code**

In [132]:
student.submit_task(globals(), task_id="task_02");

## Task 3: Fill in missing data

Fill in the missing data in the **price** column with the following procedure:

- compute the mean and std of the avaialable prices
- sample from a normal distribution with the computed mean and std (see [`np.random.normal`])(https://numpy.org/doc/stable/reference/random/generated/numpy.random.normal.html), as many samples as missing data
- substitute the missing values with the samples

For instance, for this input dataframe:

             price  category  margin
    itemid                          
    18922      NaN       1.0   10.32
    69500   121.25       1.0   10.22
    76442    90.25       1.0   12.60
    33863   106.51       0.0   10.26
    15904    95.87       1.0   11.51
    41946   103.47       2.0    9.85
    85451    93.08       2.0    9.56
    70028   116.68       1.0    9.11
    26860      NaN       2.0    9.71
    12807    91.48       0.0    9.77
    
your solution might be similar to this (not exactly the same as you will be sampling data):


                 price  category  margin
    itemid                              
    18922    97.441188       1.0   10.32
    69500   121.250000       1.0   10.22
    76442    90.250000       1.0   12.60
    33863   106.510000       0.0   10.26
    15904    95.870000       1.0   11.51
    41946   103.470000       2.0    9.85
    85451    93.080000       2.0    9.56
    70028   116.680000       1.0    9.11
    26860   103.294843       2.0    9.71
    12807    91.480000       0.0    9.77
    
    
**WARN**: your function must not modify the original dataframe, make a copy of the input dataframe, fill the values in the copy and return it.

**HINT**: use the [`isna()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.isna.html) method of a dataframe or a series.

In [ ]:
d = create_df(missing=True)
d

In [ ]:
def fillna(df):
    # make sure to make a copy in case you modify the original df
    df = df.copy()

    ... # YOUR CODE HERE
    result = ...
    return result

**check manually your code**

In [ ]:
d = create_df(missing=True)
d

In [ ]:
fillna(d)

**submit your code**

In [ ]:
student.submit_task(globals(), task_id="task_03");